Update EventDispatcher::dispatchEvent() to take in a std::span instead of a const Vector&

cdumez opened the pull request:
#### 4468b14aea09a5549dc4e05cda0c94ed09542ef1
<pre>
Update EventDispatcher::dispatchEvent() to take in a std::span instead of a const Vector&amp;
<a href="https://bugs.webkit.org/show_bug.cgi?id=293097">https://bugs.webkit.org/show_bug.cgi?id=293097</a>

Reviewed by Darin Adler.

Update EventDispatcher::dispatchEvent() to take in a std::span instead of a const Vector&amp;.
This avoids having to construct a vector at some of the call sites.

* Source/WebCore/Modules/indexeddb/IDBRequest.cpp:
(WebCore::IDBRequest::dispatchEvent):
* Source/WebCore/Modules/indexeddb/IDBTransaction.cpp:
(WebCore::IDBTransaction::dispatchEvent):
* Source/WebCore/dom/EventDispatcher.cpp:
(WebCore::dispatchEventWithType):
(WebCore::EventDispatcher::dispatchEvent):
* Source/WebCore/dom/EventDispatcher.h:
* Source/WebCore/dom/EventPath.cpp:
(WebCore::EventPath::EventPath):
* Source/WebCore/dom/EventPath.h:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::dispatchEvent):

Canonical link: <a href="https://commits.webkit.org/295022@main">https://commits.webkit.org/295022@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e6bd4a62ac9db9b90d41448126dfdf2c840dd460

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103820 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23522 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109013 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54472 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105860 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23875 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32068 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78880 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106826 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18540 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93654 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59206 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11703 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53848 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88102 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11767 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/111400 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; compiling") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30976 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22846 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/111400 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; compiling") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31338 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89855 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/111400 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; compiling") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22287 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32421 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10148 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25340 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30904 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36216 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/30698 "Build was cancelled. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Running apply-patch; Checked out pull request; compiling") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34033 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/32259 "Build was cancelled. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (approved):
This all seems fine, although a bit wordy with having to write `span` everywhere. I also wonder if we could have this take references instead of pointers since there’s no need to support nulls, but maybe that would make things worse instead of better

geoffreygaren commented on Source/WebCore/html/OffscreenCanvas.cpp:
```diff
@@ -462,7 +462,7 @@ void OffscreenCanvas::queueTaskKeepingObjectAlive(TaskSource source, Function<vo
 void OffscreenCanvas::dispatchEvent(Event& event)
 {
-    EventDispatcher::dispatchEvent({ this }, event);
+    EventDispatcher::dispatchEvent(std::initializer_list<EventTarget*>({ this }), event);
```
I'm curious: Why prefer std::initializer_list to std::array in cases like this?

cdumez replied:
Hmm, They are very similar. A std::initializer_list felt right because that's what we were using originally (though we used it implicitly to construct a vector and now we just use it straight up, without constructing a vector).

I think one drawback to std::array is that you have to specify the size at compile time if the type cannot be deduced:
```
std::initializer_list<EventTarget*>({ this })
// vs
std::array<EventTarget*, 1>({ this })
// vs
std::to_array<EventTarget*>({ this }) // Not sure if the playstation/windows port supports this yet. It didn't last I checked.
```

cdumez committed:
Update EventDispatcher::dispatchEvent() to take in a std::span instead of a const Vector&
https://bugs.webkit.org/show_bug.cgi?id=293097

Reviewed by Darin Adler.

Update EventDispatcher::dispatchEvent() to take in a std::span instead of a const Vector&.
This avoids having to construct a vector at some of the call sites.

* Source/WebCore/Modules/indexeddb/IDBRequest.cpp:
(WebCore::IDBRequest::dispatchEvent):
* Source/WebCore/Modules/indexeddb/IDBTransaction.cpp:
(WebCore::IDBTransaction::dispatchEvent):
* Source/WebCore/dom/EventDispatcher.cpp:
(WebCore::dispatchEventWithType):
(WebCore::EventDispatcher::dispatchEvent):
* Source/WebCore/dom/EventDispatcher.h:
* Source/WebCore/dom/EventPath.cpp:
(WebCore::EventPath::EventPath):
* Source/WebCore/dom/EventPath.h:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::dispatchEvent):

Canonical link: https://commits.webkit.org/295022@main
